txscript/engine.go

```diff
@@ -321,6 +330,72 @@ func (vm *Engine) Step() (done bool, err error) {
 			// Set stack to be the stack from first script minus the
 			// script itself
 			vm.SetStack(vm.savedFirstStack[:len(vm.savedFirstStack)-1])
+		} else if (vm.scriptIdx == 1 && vm.witness) ||
+			(vm.witness && vm.bip16 && vm.scriptIdx == 2) { // Nested P2SH.
+			// TODO(roasbeef): check that sigscript is just a push
+			vm.scriptIdx++
+
+			witness := vm.tx.TxIn[vm.txIdx].Witness
+			if vm.witnessVersion == 0 {
+				switch len(vm.witnessProgram) {
+				case 20: // P2WKH
+					// The witness stack should consist of exactly two
+					// items: the signature, and the pubkey.
+					if len(witness) != 2 {
+						// TODO(roasbeef): actual error
+						return false, fmt.Errorf("invalid witness for p2wkh")
+					}
+
+					// Now we'll resume execution as if it were a
+					// regular p2pkh transaction.
+					pkScript, err := payToPubKeyHashScript(vm.witnessProgram)
+					if err != nil {
+						return false, err
+					}
+					pops, err := parseScript(pkScript)
+					if err != nil {
+						return false, err
+					}
+
+					// Set the stack to the provided witness
+					// stack, then append the pkScript
+					// generated above as the next script to execute.
+					vm.scripts = append(vm.scripts, pops)
+					vm.SetStack(witness)
```

Comment:
This case will only be executed for p2wkh transactions. During core's execution, they start with an empty stack, then _assign_ this stack variable to point to the transaction's included witness stack (`vm.tx.TxIn[vm.txIdx.Witness`). This verification is done locally within a separate function so there was nothing on the stack before hand. Therefore I feel this sections is correct for p2wkh programs. But as you point out, it's _very important_ that we verify so. 

The witness program for nested P2SH outputs _must_ be the variant which pushes the 32-byte hash. So it will never execute this case which is reserved for p2wkh programs. To my current knowledge, when core validates nested P2SH outputs, it again calls the `VerifyWitnessProgram` function which begins with a fresh empty stack, assigning the stack to the transaction's provided witness data. Therefore once again the execution starts with solely the transaction's witness data on the stack. However, once the program execution concludes, the stack is _resized_ forcing the stack to only contain a single value. This value is then checked for "truthyness", as well as the clean stack assertion. 

My opinions of correct are drawn only from the fact that this branch currently sync successfully, and my my manual execution of p2wkh, p2wsh, and nested p2sh programs. However, of course there may be lingering edge cases this code doesn't yet address. Once I finish the lingering TODO's all add the new script, and transaction tests to ensure the code at least passes those additionally.